### PR TITLE
feat: implement gap fill with previous value

### PR DESCRIPTION
**Pull request by wolffcm**
Fixes https://github.com/influxdata/influxdb_iox/issues/6973.

### CODE REVIEW

File: iox_query/src/exec/gapfill/exec_tests.rs [hunk 1]
```diff
@@ -496,6 +496,81 @@ fn test_gapfill_simple_no_lower_bound() {
...
+            let actual = batches_to_lines(&batches);
+            insta::with_settings!({
+                description => format!("input_batch_size: {input_batch_size}, output_batch_size: {output_batch_size}"),
+            }, {
+                insta::assert_yaml_snapshot!(actual, @r###"
+                ---
+                - +----+--------------------------+----+
+                - "| g0 | time                     | a0 |"
+                - +----+--------------------------+----+
+                - "| a  | 1970-01-01T00:00:00.975Z |    |"
```
Comment (alamb) [inline iox_query/src/exec/gapfill/exec_tests.rs:552]: `a0` is NULL because the first row in the data window was `1000`, right (and so anything before is filled with nulls)?

File: iox_query/src/exec/gapfill/exec_tests.rs [hunk 2]
```diff
@@ -496,6 +496,81 @@ fn test_gapfill_simple_no_lower_bound() {
...
+                    // 975
+                    Some(1000),
+                    // 1025
+                    // 1050
+                    Some(1075),
+                    // 1100
+                    // 1125
+                ],
+                agg_cols: vec![vec![
+                    Some(10),
```
Comment (alamb) [inline iox_query/src/exec/gapfill/exec_tests.rs:529]: Would it be worth also testing if the aggregate value itself was NULL for one of the groups?

File: iox_query/src/exec/gapfill/algo.rs [hunk 3]
```diff
@@ -803,6 +980,143 @@ mod tests {
...
+            None,
+            None,
+            // 950,
+            Some(1000),
+            // 1050
+            Some(1100),
+            // 1150
+            Some(1200),
+            // 1250
+            //
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo.rs:1036]: maybe a left over `//`

**Review by alamb — APPROVED** [review]

File: iox_query/src/exec/gapfill/algo.rs (hunk 3 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/algo.rs:1036]: I added this on purpose because otherwise `rustfmt` will try to put the comments after each element on the same line, which makes this harder to read.

File: iox_query/src/exec/gapfill/exec_tests.rs (hunk 2 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/exec_tests.rs:529]: Yes, I will update the test.

File: iox_query/src/exec/gapfill/exec_tests.rs (hunk 1 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/exec_tests.rs:552]: Yes, you have it right. `LOCF` has a `prev` parameter to let the user select what to put here, but InfluxQL does not have anything like that, so I am deferring that until later.
